File: elixir/rna-transcription/lib/rna_transcription.ex
defmodule RnaTranscription do
  @doc """
  Transcribes a character list representing DNA nucleotides to RNA

  ## Examples

    iex> RnaTranscription.to_rna(~c"ACTG")
    ~c"UGAC"
  """
  @spec to_rna([char]) :: [char]
  def to_rna(dna) do
    dna
    |> Enum.map(&rna_complement/1)
  end

  @spec rna_complement(char) :: char
  defp rna_complement(char)

  defp rna_complement(?G), do: ?C
  defp rna_complement(?C), do: ?G
  defp rna_complement(?T), do: ?A
  defp rna_complement(?A), do: ?U
end

File: c/rna-transcription/rna_transcription.c
#include "rna_transcription.h"

char rna_complement(const char nucleotide);

char *to_rna(const char *dna) {
  char *translated = (char *)calloc(strlen(dna), sizeof(char));
  for (size_t i = 0; i < strlen(dna); i++) {
    translated[i] = rna_complement(dna[i]);
  }
  return translated;
}

char rna_complement(const char nucleotide) {
  switch (nucleotide) {
  case 'A':
    return 'U';
  case 'T':
    return 'A';
  case 'C':
    return 'G';
  case 'G':
    return 'C';
  default:
    return ' ';
  }
}